src/internal/pfsdb/commit_provenance.go

```diff
@@ -166,3 +166,123 @@ func addCommitProvenance(tx *pachsql.Tx, from, to int) error {
 	_, err := tx.Exec(stmt, from, to)
 	return errors.Wrapf(err, "add commit provenance")
 }
+
+type commitWithDepth struct {
+	Commit
+	Depth uint32 `db:"depth"`
+}
+
+// GetCommitWithIDProvenance returns the full provenance of a branch, i.e. all branches that it either directly or transitively depends on.
+// It accepts a GraphOpt option.
+func GetCommitWithIDProvenance(ctx context.Context, ext sqlx.ExtContext, startId CommitID, opt ...GraphOpt) ([]*CommitWithID, error) {
+	commitsWithDepth, err := getCommitProvenanceWithDepth(ctx, ext, startId, opt...)
+	if err != nil {
+		return nil, errors.Wrap(err, "get commit with id provenance")
+	}
+	return commitWithDepthToCommitWithID(ctx, ext, commitsWithDepth)
+}
+
+func getCommitProvenanceWithDepth(ctx context.Context, ext sqlx.ExtContext, commitId CommitID, opt ...GraphOpt) ([]commitWithDepth, error) {
+	option := defaultOption()
+	if len(opt) > 0 {
+		option.Merge(opt[0])
+	}
+	var commits []commitWithDepth
+	if err := sqlx.SelectContext(ctx, ext, &commits, `
+		WITH RECURSIVE prov(from_id, to_id) AS (
+		    SELECT from_id, to_id, 1 as depth
+		    FROM pfs.commit_provenance
+		    WHERE from_id = $1
+		  UNION ALL
+		    SELECT cp.from_id, cp.to_id, depth+1
+		    FROM prov p
+		    JOIN pfs.commit_provenance cp ON cp.from_id = p.to_id
+		)
+		SELECT DISTINCT`+commitFields+","+
+		`	MIN(depth) as "depth"
+		FROM prov p
+		JOIN pfs.commits commit ON p.to_id = commit.int_id
+		JOIN pfs.repos repo ON commit.repo_id = repo.id
+		JOIN core.projects project ON repo.project_id = project.id
+		LEFT JOIN pfs.branches branch ON commit.branch_id = branch.id
+		GROUP BY`+commitFieldsGroupBy+`
```

Comment:
Is this needed for the min depth?

Reply:
Yup

Reply:
The min depth for a node should be the only one encountered in the traversal due to the distinct, but I don't think we should return the depth anyways. We should just return the commits / branches up to a certain depth. The testing should know based on how it structured the graph which commits / branches are expected for a particular depth.